protocolapi/block/api.go

```diff
@@ -0,0 +1,160 @@
+	startMiningFunc func(context.Context) error,
+	stopMiningfunc func(context.Context),
+	syncer chain.Syncer,
+	signer consensus.TicketSigner,
+) API {
+	newDeps := apiDeps{
+		addNewBlockFunc:   addNewBlockFunc,
+		blockStore:        bstore,
+		cBorStore:         cborStore,
+		onlineStore:       onlineStore,
+		chainReader:       chainReader,
+		consensusProtocol: con,
+		blockTime:         blockTime,
+		mineDelay:         blockMineDelay,
+		msgPool:           msgPool,
+		nodePorcelain:     nodePorc,
+		powerTable:        ptv,
+		startMiningFunc:   startMiningFunc,
+		stopMiningFunc:    stopMiningfunc,
+		syncer:            syncer,
+		ticketSigner:      signer,
+	}
+
+	return API{apiDeps: newDeps}
+}
+
+// MiningOnce mines a single block in the given context, and returns the new block.
+func (a *API) MiningOnce(ctx context.Context) (*types.Block, error) {
+
+	getStateByKey := func(ctx context.Context, tsKey string) (state.Tree, error) {
+		tsas, err := a.chainReader.GetTipSetAndState(ctx, tsKey)
+		if err != nil {
+			return nil, err
+		}
+		return state.LoadStateTree(ctx, a.cBorStore, tsas.TipSetStateRoot, builtin.Actors)
+	}
+
+	getState := func(ctx context.Context, ts types.TipSet) (state.Tree, error) {
+		return getStateByKey(ctx, ts.String())
+	}
+
+	getWeight := func(ctx context.Context, ts types.TipSet) (uint64, error) {
+		parent, err := ts.Parents()
+		if err != nil {
+			return uint64(0), err
+		}
+		if parent.Len() == 0 {
+			return a.consensusProtocol.Weight(ctx, ts, nil)
+		}
+		pSt, err := getStateByKey(ctx, parent.String())
+		if err != nil {
+			return uint64(0), err
+		}
+		return a.consensusProtocol.Weight(ctx, ts, pSt)
+	}
+
+	minerAddrIf, err := a.nodePorcelain.ConfigGet("mining.minerAddress")
+	if err != nil {
+		return nil, err
+	}
+	minerAddr := minerAddrIf.(address.Address)
+	minerOwnerAddr, err := a.nodePorcelain.MinerGetOwnerAddress(ctx, minerAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	minerPubKey, err := a.nodePorcelain.MinerGetKey(ctx, minerAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	getAncestors := func(ctx context.Context, ts types.TipSet, newBlockHeight *types.BlockHeight) ([]types.TipSet, error) {
+		return chain.GetRecentAncestors(ctx, ts, a.chainReader, newBlockHeight, consensus.AncestorRoundsNeeded, sampling.LookbackParameter)
+	}
+
+	worker := mining.NewDefaultWorker(
+		a.msgPool, getState, getWeight, getAncestors, consensus.NewDefaultProcessor(),
+		a.powerTable, a.blockStore, a.cBorStore, minerAddr, minerOwnerAddr, minerPubKey,
+		a.ticketSigner, a.blockTime)
+
+	ts := a.chainReader.Head()
+	res, err := mining.MineOnce(ctx, worker, a.mineDelay, ts)
+	if err != nil {
+		return nil, err
+	}
+	if res.Err != nil {
+		return nil, res.Err
+	}
+
+	if err := a.addNewBlockFunc(ctx, res.NewBlock); err != nil {
+		return nil, err
+	}
+
+	return res.NewBlock, nil
+}
+
+// MiningStart calls the node's StartMining function
+func (a *API) MiningStart(ctx context.Context) error {
```

Comment:
i think we should pass the mining node's mining worker to the api and have the api call directly into it. probably all the bookkeeping stuff strewn throughout the node that keeps track of mining state should move here as well. can you add this to the list of cleanups?

Reply:
#2277